imports/server/gdriveActivityFetcher.ts

```diff
@@ -0,0 +1,171 @@
+import MeteorUsers from '../lib/models/MeteorUsers';
+import roundedTime from '../lib/roundedTime';
+import GoogleClient from './googleClientRefresher';
+import ignoringDuplicateKeyErrors from './ignoringDuplicateKeyErrors';
+import DriveActivityLatests from './models/DriveActivityLatests';
+import Locks, { PREEMPT_TIMEOUT } from './models/Locks';
+
+async function recordDriveChanges(ts: Date, fileIds: string[], googleAccountIds: string[]) {
+  const time = roundedTime(ACTIVITY_GRANULARITY, ts);
+
+  // In all likelihood, we will only have one of each of these, but for
+  // completeness we'll record the full cartesian product
+  for await (const fileId of fileIds) {
+    const document = await Documents.findOneAsync({ 'value.id': fileId });
+    if (!document) {
+      // eslint-disable-next-line no-continue
+      continue;
+    }
+
+    for await (const googleAccountId of googleAccountIds) {
+      // if we can't look up the account ID, we'll record as user=undefined
+      // which we'll count as a seprate user
+      const user = await MeteorUsers.findOneAsync({ googleAccountId, hunts: document.hunt });
+
+      await ignoringDuplicateKeyErrors(async () => {
+        await DocumentActivities.insertAsync({
+          ts: time,
+          document: document._id,
+          hunt: document.hunt,
+          puzzle: document.puzzle,
+          user: user?._id,
+        });
+      });
+    }
+  }
+}
+
+async function fetchDriveActivity() {
+  const { driveactivity } = GoogleClient;
+  if (!driveactivity) {
+    return;
+  }
+
+  // Don't fetch history that's older than what we'd display
+  const previousTimestamp = Math.max(
+    (await DriveActivityLatests.findOneAsync('default'))?.ts.getTime() ?? 0,
+    Date.now() - ACTIVITY_GRANULARITY * ACTIVITY_SEGMENTS,
+  );
+
+  // Build in some buffer by starting 5 minutes before the latest timestamp
+  // we've previously seen (our unique constraints will dedup any overlap)
+  const filter = `time > ${previousTimestamp - 5 * 60 * 1000}`;
+
+  let pageToken: string | undefined;
+  let latestTimestamp = previousTimestamp;
+  do {
+    const resp = await driveactivity.activity.query({
+      requestBody: {
+        pageToken,
+        filter,
+      },
+    });
+    pageToken = resp.data.nextPageToken ?? undefined;
+
+    if (resp.data.activities) {
+      // Accumulate a promise that resolves to the latest timestamp we've seen
+      latestTimestamp = await resp.data.activities.reduce(async (p, activity) => {
+        const previousLatestTimestamp = await p;
+
+        // See if this is a document edit action
+        if (!activity.actions?.some((action) => action.detail?.edit)) {
+          return previousLatestTimestamp;
+        }
+
+        if (!activity.timestamp || !activity.targets || !activity.actors) {
+          return previousLatestTimestamp;
+        }
+
+        const ts = new Date(activity.timestamp);
+
+        const documentIds = [...activity.targets.reduce<Set<string>>((acc, target) => {
```

Comment:
It took me a couple reads through this before I realized that in all likelihood any given `activity` will have just 1 document id or just 1 actor id, and we're just handling both possible aggregations (and if the power set really happens, then we'll handle that correctly too).  I might add a comment here explaining that we mostly expect one or the other to be a single element, but this handles all the possibilities neatly.